libs/auth/react/src/lib/Authenticator/Authenticator.tsx

```diff
@@ -0,0 +1,158 @@
+
+interface Props {
+  /**
+   * If true, Authenticator automatically starts login flow and does not render children until user is fully logged in.
+   * If false, children are responsible for rendering a login button and loading indicator.
+   * Default: true
+   */
+  autoLogin?: boolean
+}
+
+export const Authenticator: FC<Props> = ({ children, autoLogin = true }) => {
+  const history = useHistory()
+  const reducerInstance = useReducer(reducer, initialState)
+  const [state, dispatch] = reducerInstance
+  const userManager = getUserManager()
+  const authSettings = getAuthSettings()
+
+  const signIn = useCallback(
+    async function signIn() {
+      dispatch({
+        type: ActionType.SIGNIN_START,
+      })
+      return userManager.signinRedirect({
+        state: history.location.pathname + history.location.search,
+      })
+      // Nothing more happens here since browser will redirect to IDS.
+    },
+    [dispatch, userManager, history],
+  )
+
+  const signInSilent = useCallback(
+    async function signInSilent() {
+      let user = null
+      dispatch({
+        type: ActionType.SIGNIN_START,
+      })
+      try {
+        user = await userManager.signinSilent()
+        dispatch({ type: ActionType.SIGNIN_SUCCESS, payload: user })
+      } catch (error) {
+        dispatch({ type: ActionType.SIGNIN_FAILURE })
+      }
+
+      return user
+    },
+    [userManager, dispatch],
+  )
+
+  const signOut = useCallback(
+    async function signOut() {
+      dispatch({
+        type: ActionType.LOG_OUT,
+      })
+      await userManager.signoutRedirect()
+    },
+    [userManager, dispatch],
+  )
+
+  const checkLogin = useCallback(
+    async function checkLogin() {
+      dispatch({
+        type: ActionType.SIGNIN_START,
+      })
+
+      const storedUser = await userManager.getUser()
+
+      // Check expiry.
+      if (storedUser && !storedUser.expired) {
+        dispatch({
+          type: ActionType.SIGNIN_SUCCESS,
+          payload: storedUser,
+        })
+      } else if (autoLogin) {
+        // If we find a user in SessionStorage, there's a fine chance that
+        // it's just an expired token and we can silently log in.
+        if (storedUser && (await signInSilent())) {
+          return
+        }
+
+        // If all else fails, redirect to the login page.
+        await signIn()
+      } else {
+        // When not performing autologin, silently check if there's an IDP session.
+        await signInSilent()
+      }
+    },
+    [userManager, dispatch, signIn, signInSilent],
+  )
+
+  useEffect(() => {
+    // This is raised when a new user state has been loaded with a silent login.
+    const userLoaded = (user: User) => {
+      dispatch({
+        type: ActionType.USER_LOADED,
+        payload: user,
+      })
+    }
+
+    // This is raised when the user is signed out of the IDP.
+    const userSignedOut = () => {
+      dispatch({
+        type: ActionType.LOG_OUT,
+      })
+      if (autoLogin) {
```

Comment:
Calling signIn here to redirect the user back to the IDS after logout?

Reply:
Yeah. If the user initiates a logout themself, they are sent (by the IDS) to the `post_logout_redirect_uri` (usually island.is front page).

But if the IDS session expires (1 hour) or if the user logs out in another tab, we call signin here to redirect this tab to the IDS login screen. This means:

* any sensitive information is hidden from the screen.
* the user is aware that the session is over, and doesn't try to submit changes with an expired session.
* the user can finish logging in and end up back where they were.

We'll see how this works and "feels" in practice, e.g. with delegations.

This is similar to google btw, though they take things a step further, so if I have 10 inactive tabs, all showing the login screen, it's enough for me to login to one of them, and all will then go back to where they were. That might be a cool addition here.